### ISSUE TITLE
Error protecting file in service application

### ISSUE THREAD

**Issue by OleksiiD** [MicrosoftDocs/Azure-RMSDocs#955]
Hi, can someone help me with the following issue?

**Scenario**. I have a Windows Service running on an Azure VM. Service receives files, modifies them in some way (let's assume that it adds custom properties to Word files) and uses MIP SDK to protect them with template ID.

**Issue**.  `IFileHandler.SetProtection(string)+CommitAsync(...)` fails with the following exception:

_One or more errors occurred.\n\nServiceDiscoveryHelper::GetServiceDetails - Cannot compute domain: license domains, identity, and cloud endpoint base URL are all empty, correlationId:[9add32ba-0cb7-4d31-b9d8-0000b7c694a4]_

**Other info**
- `RemoveProtection()+CommitAsync(...)` work fine.
- I registered application in Azure Active Directory tenant.
  - Generated secret: `<CLIENT_SECRET>`.
  - Granted the following permissions
    - https://api.mipwebservice.com/InformationProtectionPolicy.Read.All
    - https://psor.o365syncservice.com/UnifiedPolicy.Tenant.Read
    - https://aadrm.com/Content.SuperUser
    - https://aadrm.com/Content.Writer
    - https://aadrm.com/Content.DelegatedWriter
    - https://aadrm.com/Content.DelegatedReader
- IAuthDelegate implementation
  - uses ADAL to get access token using `client_credentials` authentication flow, because there is no interacting user (my app is service).
  - I do not whether I have to use `identity` parameter in `client_credentials` flow.

**Main Code Snippet**
```c#
MIP.Initialize(MipComponent.File);

var appInfo = new ApplicationInfo{
    ApplicationId = ConfigurationManager.AppSettings["AppPrincipalId"],
    ApplicationName = "App name",
    ApplicationVersion = "1.0.0",
};

var authDelegate = new AuthDelegateImplementation(appInfo);
var fileProfileSettings = new FileProfileSettings("mip_data", false,
    authDelegate, new ConsentDelegateImplementation(), appInfo, LogLevel.Trace);

var fileProfile = MIP.LoadFileProfileAsync(fileProfileSettings).Result;
var engineSettings = new FileEngineSettings("engine-id", "", "en-US"){
    Identity = new Identity($"{appInfo.ApplicationId}@<TENANT-NAME>"){
        DelegatedEmail = "<OWNER>@<TENANT-NAME>",
    },
};
var fileEngine = fileProfile.AddEngineAsync(engineSettings).Result;
var fileHandler = fileEngine.CreateFileHandlerAsync("c:\\sample.docx", "0", true).Result;
fileHandler.SetProtection(new ProtectionDescriptor("<TEMPLATE-ID>"));
var success = fileHandler.CommitAsync("c:\\encrypted.docx").Result;
```
**AuthDelegateImplementation**
```c#
public string AcquireToken(Identity identity, string authority, string resource)
    var authContext = new AuthenticationContext(authority + "/" + "<TENANT_ID>");
    var clientCredential = new ClientCredential("<CLENT_ID>", "<CLIENT_SECRET>");
    var res = await authContext.AcquireTokenAsync(resource, clientCredential);
    return res.AccessToken;
}
```
**ConsentDelegateImplementation**
```c#
public class ConsentDelegateImplementation : IConsentDelegate {
    public Consent GetUserConsent(string url) {
        return Consent.Accept;
    }
}
```

**Comment by OleksiiD** [comment]
It seams that after testing MIP's local persistent state (see FileProfileSettings.Path property when UseInMemoryStorage is flase) was corrupted. After removing "mip_data" folder issue  disappeared.

**Comment by garima2510** [comment]
I am facing the same issue while trying to SetLabel on PDF file stream coming from SharePoint. Removing the mip folder did not help. Any idea please?

**Comment by OleksiiD** [comment]
@garima2510 It is hard to tell w/o code. Did you try using in-memory storage?

**Comment by garima2510** [comment]
Yes, I finally made it work passing a valid admin account email id while creating the File Engine Settings (line 40). The complete code is present here: [MIP SDK with SharePoint](https://mysharepointlearnings.wordpress.com/2019/07/26/using-mip-sdk-in-sharepoint/)

My code is almost same as yours and even I am using client credentials flow. Problem is below line:

var engineSettings = new FileEngineSettings("engine-id", "", "en-US"){
    Identity = new Identity($"{appInfo.ApplicationId}@<TENANT-NAME>"){
        DelegatedEmail = "<OWNER>@<TENANT-NAME>",
    },
};

does not work for me. I have to pass owner@tenant-name as identity email and blank in delegated email for this to function properly. 

Apart from this, the PDF File has label set using identity of Azure AD app and I am not able to open it using AIP viewer. It says access denied even though it has access. Did you face this issue?

**Comment by OleksiiD** [comment]
Yes, I did face with this problem. After setting label with protection, client identity became an owner of the document.

To fix this you must set DelegatedEmail to the email address of the user (UPN) on whose behalf you are going to set this label.

**Comment by garima2510** [comment]
Thanks for taking out time to reply but in my case if I pass UPN (email id) of user in Delegated Email it gives another error:

The service didn't accept the auth token. AuthChallenge:[Bearer resource="https://aadrm.com", realm="---", authorization="https://login.windows.net/tenantid/oauth2/authorize"], correlationId:[2d3c9fa2-cbe7-4350-9172-00006e9a41ad]

**Comment by darioairoldi** [comment]
"The service didn't accept the auth token" is returned because the token was acquired for the wrong scope.

- when **AcquireToken** is called with resource **"https://syncservice.o365syncservice.com/"** you must call 
   app.AcquireTokenSilent(new[]{ **"https://psor.o365syncservice.com/UnifiedPolicy.User.Read"** }, firstAccount).ExecuteAsync();

- when **AcquireToken** is called with resource **"https://aadrm.com"** you must call 
   app.AcquireTokenSilent(new[]{ **"https://aadrm.com/user_impersonation"** }, firstAccount).ExecuteAsync();

**Comment by vimal-amaldas-git** [comment]
Based on the resource type, the scope has to be switched dynamically in the acquire token method.
The initial File engine is accepting only "https://psor.o365syncservice.com/UnifiedPolicy.User.Read" as scope
 var engineSettings = new FileEngineSettings("user1@tenant.com", authDelegate, "", "en-US");

For any other action such as creating File Handler "https://aadrm.com/user_impersonation" is the accepted scope

var fileHandler = fileEngine.CreateFileHandlerAsync("c:\\sample.docx", "0", true).Result;

So basically, acquireToken method should contain statements to switch scope based on resource something like

AcquireToken(string authority,string resource,string claims)
{
...
if(resource=="https://psor.o365syncservice.com")
    scope ="https://psor.o365syncservice.com/UnifiedPolicy.User.Read";
else
   scope ="https://aadrm.com/user_impersonation";

}
